contracts/btc-finality/src/finality.rs

```diff
@@ -894,52 +894,80 @@ pub fn query_fps_by_total_active_sats(
     Ok(res.fps)
 }
 
-/// Distributes rewards to finality providers who are in the active set at `height`.
-pub fn distribute_rewards_fps(deps: &mut DepsMut, env: &Env) -> Result<(), ContractError> {
-    // Try to use the finality provider set at the previous height
-    let active_fps = get_power_table_at_height(deps.storage, env.block.height - 1)?;
-    // Short-circuit if there are no active finality providers
-    if active_fps.is_empty() {
-        return Ok(());
-    }
-    // Get the voting power of the active FPS
-    let total_voting_power = active_fps
-        .values()
-        .map(|power| *power as u128)
-        .sum::<u128>();
-    // Short-circuit if the total voting power is zero
-    if total_voting_power == 0 {
-        return Ok(());
-    }
-    // Get the rewards to distribute (bank balance of the finality contract, minus previously / already distributed rewards
-    // (pending to be sent to Babylon on an epoch boundary))
-    let total_pending_rewards = TOTAL_PENDING_REWARDS.load(deps.storage)?;
+/// Distributes rewards in a given height range based on vote participation weighted by voting power.
+/// This approach counts the total votes each FP made across the specified range, weights them by
+/// the FP's voting power at each height they voted, and distributes rewards proportionally.
+pub fn distribute_rewards_in_range(
+    deps: &mut DepsMut,
+    env: &Env,
+    start_height: u64,
+    end_height: u64,
+) -> Result<(), ContractError> {
     let cfg = CONFIG.load(deps.storage)?;
-    let rewards_amount = deps
+
+    // Get current balance of the finality contract (total rewards to distribute)
+    let current_balance = deps
         .querier
         .query_balance(env.contract.address.clone(), cfg.denom)?
-        .amount
-        .saturating_sub(total_pending_rewards);
-    // Short-circuit if there are no rewards to distribute
-    if rewards_amount.is_zero() {
+        .amount;
```

Comment:
We probably need to leverage the reward storage ([ref](https://github.com/babylonlabs-io/cosmos-bsn-contracts/blob/86b9f83a1eb1d63cb955af4130b600689044649e/contracts/btc-finality/src/state/finality.rs#L64-L69)) for getting the reward to be distributed. This was to prevent the contracts from distributing random tokens that contracts receive from random people

Reply:
I don't understand. We have to obtain the rewards from the balance of the finality contract even though it has balance from random reason, no?

Reply:
The balance needs to be fetched from the `TOTAL_PENDING_REWARDS` storage, rather than the balance. This is because `TOTAL_PENDING_REWARDS` is strictly the value that the BSN allocates to the BTC stakers / FPs, and it might not be equal to the contract balance because the contract can receive tokens from anyone.

Reply:
I feel more confused. `TOTAL_PENDING_REWARDS` only keeps track of the rewards to send to Babylon genesis. We still need query the contracts for balance to distribute rewards. See previous impl https://github.com/babylonlabs-io/cosmos-bsn-contracts/blob/3b1af6c09d9dfec8705ff99d33b01ec8c32925c0/contracts/btc-finality/src/finality.rs#L915. In fact, we don't need `TOTAL_PENDING_REWARDS` anymore as we do not need it anymore as we do rewards dist per epoch now

Reply:
Removed it in [c5fb579](https://github.com/babylonlabs-io/cosmos-bsn-contracts/pull/291/commits/c5fb5790d7298dc54821056606641fad9c4f2295)